### PR TITLE
Microsoft Application Insights output plugin

### PR DESCRIPTION
**Pull request by karolz-ms**
This PR adds an output plugin that sends data to Microsoft Application Insights https://docs.microsoft.com/azure/application-insights/

Hopefully this addition aligns with telegraf roadmap, but if there are concerns, we (here at Microsoft) are open to other suggestions--thanks!

### CODE REVIEW

**Comment by karolz-ms** [comment]
@brahmnes @yantang-msft @jjjordanmsft Could you please take a peek? TIA!

File: plugins/outputs/application_insights/application_insights.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+		a.diagMsgListener.Remove()
+	}
+
+	return nil
+}
+
+func init() {
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			DiagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: Is there a way to disable the diagMsgListener?
I can't find the doc about configuration, but I assume user can only specify simple data type in the configuration, thus the diagnostics listener is always created?

File: plugins/outputs/application_insights/application_insights.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+func (a *ApplicationInsights) Connect() error {
+	if a.Timeout.Duration == 0 {
+		a.Timeout = internal.Duration{Duration: time.Second * 5}
+	}
+
+	if a.Transmitter == nil {
+		a.useDefaultTransmitter()
+	}
+
+	if a.Transmitter != nil && a.DiagMsgSubscriber != nil {
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: I feel like a boolean configuration like "developer_mode" is simpler than this extra interface and implementation.

File: plugins/outputs/application_insights/application_insights.go [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+			a.Transmitter.Track(telemetry)
+		}
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Close() error {
+	if a.Transmitter == nil {
+		return nil
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: nit: The current logic won't create diagMsgListener if Transmitter is null, but still feel safer to always check diagMsgListener before return.
How about
```
if a.Transmitter != nil {
    select {....}
}
// check and remove diaMsgListener
```

File: plugins/outputs/application_insights/application_insights.go [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+		}
+
+		return int(v), nil
+	}
+
+	return 0.0, fmt.Errorf("[%s] cannot be converted to an int value", value)
+}
+
+func logOutputMsg(format string, v ...interface{}) {
+	log.Printf("Output [application_insights] "+format, v...)
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: nit: space between '+'

File: plugins/outputs/application_insights/application_insights.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+				return err
+			}
+
+			*telemetryProperty = metricValue
+
+			if usedFields != nil {
+				*usedFields = append(*usedFields, fieldName)
+			}
+
+			return nil
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: Is this return statement necessary?

File: plugins/outputs/application_insights/application_insights.go [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+func setFloat64TelemetryProperty(
+	telemetryProperty *float64,
+	candidateFields []string,
+	metric telegraf.Metric,
+	usedFields *[]string) error {
+
+	for _, fieldName := range candidateFields {
+		if metric.HasField(fieldName) {
+			metricValue, err := toFloat64(metric.Fields()[fieldName])
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: What if the fieldName afterwards can be converted? Do we still want to return err?

File: plugins/outputs/application_insights/application_insights.go [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+			}
+			suffix++
+		}
+		properties[name+"_"+strconv.Itoa(suffix)] = value
+	}
+}
+
+func toFloat64(value interface{}) (float64, error) {
+	switch v := value.(type) {
+	case int:
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: Can use list in the case statement
case int, int8, int16...:

File: plugins/outputs/application_insights/application_insights.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+		return v, nil
+	}
+
+	return 0.0, fmt.Errorf("[%s] cannot be converted to a float64 value", value)
+}
+
+func toInt(value interface{}) (int, error) {
+	if !is32BitChecked {
+		is32BitChecked = true
+		i := int(-1)
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: What do you think of using sizeof to check 32bit? That is more straightforward to me.

File: plugins/outputs/application_insights/application_insights.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+	telemetry.Properties = metric.Tags()
+	telemetry.Timestamp = metric.Time()
+
+	err := setFloat64TelemetryProperty(&(telemetry.Value), []string{"value"}, metric, nil)
+	if err == nil {
+		addRemainingMeasurements(telemetry, metric, []string{"value"})
+	} else {
+		// Either the metric does not have a "value" field, or it was not convertable to float64
+		// Heuristics: if the metric has just one field that is convertible to float64, take it as value
+		// Otherwise give up
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: I think the "value" field is just a special case. And the general case is metric with multiple dimensions (fields), and we shouldn't give up in this case.
For example, I can have a metric like
{
    metricName: "cpu",
    fields: [
        "Percent_Idle_Time": 0.3,
        "Percent_User_Time": 0.5
    ]
}
and we can generate multiple AI telemetries from this metric, with metric names cpu_Percent_Idle_Time, cpu_Percent_User_Time

File: plugins/outputs/application_insights/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# Application Insights Output Plugin
+
+This plugin writes telegraf metrics to Auzre Application Insights
```
Comment (jwilder) [inline plugins/outputs/application_insights/README.md]: Azure

File: plugins/outputs/application_insights/application_insights.go (hunk 9 above)
Comment (jwilder) [reply plugins/outputs/application_insights/application_insights.go]: @yantang-msft That's correct.  A measurement (in telegraf/influx) could have multiple fields.  

The logic here would would only pick up a measurement that happens to have a `value` field and is convertible to a `float64`... dropping all other fields which is unlikely what is intended.

File: plugins/outputs/application_insights/application_insights.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+
+func addRemainingMeasurements(telemetry appinsights.Telemetry, metric telegraf.Metric, usedFields []string) {
+	properties := telemetry.GetProperties()
+
+	for fieldName, fieldValue := range metric.Fields() {
+		if contains(usedFields, fieldName) {
+			continue
+		}
+
+		float64Value, err := toFloat64(fieldValue)
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: Since 'properties' is a string to string map, is it still necessary to require the fieldValue to be numeric? Although I believe the value is always a numeric. :)

File: plugins/outputs/application_insights/application_insights.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+	return false
+}
+
+func addUniquelyNamedProperty(properties map[string]string, name string, value string) {
+	if _, found := properties[name]; !found {
+		properties[name] = value
+	} else {
+		suffix := 1
+		for {
+			_, found = properties[name+"_"+strconv.Itoa(suffix)]
```
Comment (yantang-msft) [inline plugins/outputs/application_insights/application_insights.go]: Looks like you intended not to having space with string concatenation.

Comment (yantang-msft) [reply plugins/outputs/application_insights/application_insights.go]: Is it possible to use the same if statement like above to simplify this? i.e.,
if _, found = properties[...]; !found {
    break;
}

File: plugins/outputs/application_insights/application_insights.go (hunk 1 above)
Comment (jjjordanmsft) [reply plugins/outputs/application_insights/application_insights.go]: It's a function callback, so you could disable it in your implementation.  The SDK provides only Add/Remove.

File: plugins/outputs/application_insights/application_insights.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+	err = setIntTelemetryProperty(&(telemetry.Count), []string{"count", "samples"}, metric, &usedFields)
+	if err != nil {
+		return nil
+	}
+
+	// We attempt to set Min, Max, StdDev and Variance fields but do not really care if they are not present
+	setFloat64TelemetryProperty(&(telemetry.Min), []string{"min"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Max), []string{"max"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.StdDev), []string{"stddev"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Variance), []string{"variance"}, metric, &usedFields)
```
Comment (jjjordanmsft) [inline plugins/outputs/application_insights/application_insights.go]: Is there a reason to set both?  The SDK will only use one or the other, with stddev taking precedence.

File: plugins/outputs/application_insights/application_insights.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+		float64Value, err := toFloat64(fieldValue)
+		if err != nil {
+			continue
+		}
+
+		// We would like to add the field to telemetry.Measurements here
+		// But because of issue https://github.com/Microsoft/ApplicationInsights-Go/issues/12
+		// this is currently impossible. So we are just going to add the measurement as a property instead.
+		// This is what we want to be able to do, eventually:
+		//     telemetry.GetMeasurements()[fieldName] = float64Value
```
Comment (jjjordanmsft) [inline plugins/outputs/application_insights/application_insights.go]: I assume that `telemetry` is always a metric (whether single or multiple value)?

This is unfortunate.  We lose many possible ways to analyze these measurements by stringifying them into properties (which, btw, I'm not sure you can even see except in the AA portal)

Is it really not feasible to log these as separate metrics?

File: plugins/outputs/application_insights/application_insights.go (hunk 12 above)
Comment (jjjordanmsft) [reply plugins/outputs/application_insights/application_insights.go]: @yantang-msft FYI `go fmt` would have output the string concatenation that way, and it's the accepted convention.

File: plugins/outputs/application_insights/application_insights.go (hunk 7 above)
Comment (jjjordanmsft) [reply plugins/outputs/application_insights/application_insights.go]: See [type switches](https://golang.org/doc/effective_go.html#type_switch)

File: plugins/outputs/application_insights/README.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+### Configuration
```
Comment (jjjordanmsft) [inline plugins/outputs/application_insights/README.md]: I guess I don't know a whole lot about this system, but there may be some value in specifying things like RoleName, RoleInstance, and custom properties -- all things that would just end up in the TelemetryClient.Context() object.  These help isolate regions, resource groups, and instances in our queries.  Perhaps the default (hostname in RoleInstance) is enough, but I find myself always filling in these fields for services that I monitor.

**Comment by karolz-ms** [comment]
Hey, thanks everyone for lots of good comments! The main takeaway seems that, facing with a metric with multiple fields, it is better to submit multiple metric telemetry records to AppInsights rather than squishing the fields into telemetry properties. I completely agree with this, but it will require some rework and new tests, so I am going to close this PR and submit another one that incorporates this change.

@brahmnes FYI

File: plugins/outputs/application_insights/application_insights.go (hunk 2 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: @yantang-msft Agreed, DiagMsgSubscriber was exposed for testability, but it is probably too much of a good thing. Will rework

File: plugins/outputs/application_insights/application_insights.go (hunk 3 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Agreed

File: plugins/outputs/application_insights/application_insights.go (hunk 4 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: This is how standard go code formatting formats this expression

File: plugins/outputs/application_insights/application_insights.go [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			DiagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
+		}
+	})
+}
+
+func (a *ApplicationInsights) useDefaultTransmitter() {
+	// Env variable, if present, overrides what is in the configuration
+	iKey := os.Getenv("APPINSIGHTS_INSTRUMENTATIONKEY")
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Makes sense, will change this

File: plugins/outputs/application_insights/application_insights.go (hunk 6 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: @yantang-msft Not sure I understand the question. The error is about inability to convert to float64 and has nothing to do with the field name.

File: plugins/outputs/application_insights/application_insights.go (hunk 7 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Meaning, you can't

File: plugins/outputs/application_insights/application_insights.go (hunk 8 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Sure, we can to that

File: plugins/outputs/application_insights/application_insights.go (hunk 9 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Agreed, will submit new PR to fix this

File: plugins/outputs/application_insights/README.md (hunk 10 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/README.md]: thx

File: plugins/outputs/application_insights/application_insights.go (hunk 13 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: @jjjordanmsft The logic here is that we look for both fields, and use what is available--the call will be a no-op if . I will make sure that we test for stddev after testing for variance to align with the AppInsights SDK

File: plugins/outputs/application_insights/application_insights.go (hunk 14 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: It is, will change that

File: plugins/outputs/application_insights/README.md (hunk 15 above)
Comment (karolz-ms) [reply plugins/outputs/application_insights/README.md]: @jjjordanmsft This is good point and we are having some discussions about how to set "special" properties on AppInsights telemetry. I will add you to the email thread

File: plugins/outputs/application_insights/application_insights.go [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+	}
+
+	if a.InstrumentationKey != "" {
+		a.Transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+}
+
+func canBeTreatedAsAggregate(metric telegraf.Metric) bool {
+	// If we can identify the Sum and Count properties, then we are good
+	hasSum := metric.HasField("value") || metric.HasField("sum")
```
[the remark this answers is not in the document]
Comment (yantang-msft) [reply plugins/outputs/application_insights/application_insights.go]: We hardcoded the rule of how an metric can be treated as aggregated, which would be hard to follow. What if the field name is "sample" or "sampling" instead of "samples", or "total" instead of "sum"?
Someone will have to decide whether a metric is aggreated, but I hope it's done by the input plugin, e.g., AppMetric input plugin will know the property name that means aggregated plugin, or some override plugin with more flexibility.
As an output plugin, we can just rely on this property Metric.IsAggregate.

File: plugins/outputs/application_insights/application_insights.go (hunk 6 above)
Comment (yantang-msft) [reply plugins/outputs/application_insights/application_insights.go]: I mean we have a list of candidateFields. If the first candidate field failed to convert, but the second candidate field can be converted, do we want to return as soon as there is an error or wait after all fields are checked.

Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: Ah, good point--we should continue. Thanks

File: plugins/outputs/application_insights/application_insights.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (karolz-ms) [reply plugins/outputs/application_insights/application_insights.go]: telegraf.Metric.IsAggregate() is just a flag on the (Telegraf) metric type. From Application Insights perspective, to create an AppInsights aggregate metric, we have to know at least the sum of all values and the number of samples, so telegraf.Metric.IsAggregate() alone does not help us with that task. 

Application Insights aggregate metric is a concept specific to AppInsighst, so it *has to be the AppInsights output plugin* that makes the decision whether a Telegraf metric can be interpreted as an AppInsights aggregate metric or not. We might eventually need more flexibility here in terms of how this interpretation is made, but it quickly gets complicated and I do not want to try to solve this in the first iteration.
